File: backend/app/routers/stats.py
from typing import Optional
from fastapi import APIRouter, Query

from app.core import db as db_module
from app.deps import AccountConn

router = APIRouter()


@router.get("/{account_id}/stats/kpi-values")
def get_kpi_values(
    account_conn: AccountConn,
    fecha_desde: Optional[str] = Query(None),
    fecha_hasta: Optional[str] = Query(None),
    area: Optional[str] = Query(None),
    categoria: Optional[str] = Query(None),
    tag: Optional[str] = Query(None),
):
    conn, _, _ = account_conn
    values = db_module.compute_kpi_values(conn, fecha_desde, fecha_hasta, area, categoria, tag)
    return {"values": values}


@router.get("/{account_id}/stats/monthly-summary")
def get_monthly_summary(
    account_conn: AccountConn,
    fecha_desde: Optional[str] = Query(None),
    fecha_hasta: Optional[str] = Query(None),
    area: Optional[str] = Query(None),
    categoria: Optional[str] = Query(None),
    tag: Optional[str] = Query(None),
):
    conn, _, _ = account_conn
    data = db_module.get_monthly_summary(conn, fecha_desde, fecha_hasta, area, categoria, tag)
    return {"data": data}


@router.get("/{account_id}/stats/by-category")
def get_by_category(
    account_conn: AccountConn,
    fecha_desde: Optional[str] = Query(None),
    fecha_hasta: Optional[str] = Query(None),
    area: Optional[str] = Query(None),
    tag: Optional[str] = Query(None),
    categoria: Optional[str] = Query(None),
):
    conn, _, _ = account_conn
    data = db_module.get_by_category(conn, fecha_desde, fecha_hasta, area, tag, categoria)
    return {"data": data}


@router.get("/{account_id}/stats/by-area")
def get_by_area(
    account_conn: AccountConn,
    fecha_desde: Optional[str] = Query(None),
    fecha_hasta: Optional[str] = Query(None),
    area: Optional[str] = Query(None),
    categoria: Optional[str] = Query(None),
    tag: Optional[str] = Query(None),
):
    conn, _, _ = account_conn
    data = db_module.get_by_area(conn, fecha_desde, fecha_hasta, area, categoria, tag)
    return {"data": data}


@router.get("/{account_id}/stats/monthly-by-category")
def get_monthly_by_category(
    account_conn: AccountConn,
    fecha_desde: Optional[str] = Query(None),
    fecha_hasta: Optional[str] = Query(None),
    area: Optional[str] = Query(None),
    tag: Optional[str] = Query(None),
):
    conn, _, _ = account_conn
    data = db_module.get_monthly_by_category(conn, fecha_desde, fecha_hasta, area, tag)
    return {"data": data}


@router.get("/{account_id}/stats/monthly-by-area")
def get_monthly_by_area(
    account_conn: AccountConn,
    fecha_desde: Optional[str] = Query(None),
    fecha_hasta: Optional[str] = Query(None),
    categoria: Optional[str] = Query(None),
    tag: Optional[str] = Query(None),
):
    conn, _, _ = account_conn
    data = db_module.get_monthly_by_area(conn, fecha_desde, fecha_hasta, categoria, tag)
    return {"data": data}


@router.get("/{account_id}/stats/top-comercios")
def get_top_comercios(
    account_conn: AccountConn,
    fecha_desde: Optional[str] = Query(None),
    fecha_hasta: Optional[str] = Query(None),
    limit: int = Query(10, ge=1, le=50),
    area: Optional[str] = Query(None),
    categoria: Optional[str] = Query(None),
    tag: Optional[str] = Query(None),
):
    conn, _, _ = account_conn
    data = db_module.get_top_comercios(conn, fecha_desde, fecha_hasta, limit, area, categoria, tag)
    return {"data": data}

File: frontend/src/api/stats.ts
import { apiClient } from './client'
import type {
  MonthlySummaryItem,
  CategoryBreakdown,
  AreaBreakdown,
  MonthlyCategoryItem,
  MonthlyAreaItem,
  Transaction,
} from '@/types'

interface DateRange {
  fecha_desde?: string
  fecha_hasta?: string
  area?: string
  categoria?: string
  tag?: string
  desde_ahorro?: number
}

export interface TopComercioItem {
  nombre: string
  gasto: number
  num: number
}

export const statsApi = {
  kpiValues: (accountId: number, range: DateRange = {}) =>
    apiClient
      .get<{ values: Record<number, number> }>(`/api/accounts/${accountId}/stats/kpi-values`, { params: range })
      .then((r) => r.data.values),

  monthlySummary: (accountId: number, range: DateRange = {}) =>
    apiClient
      .get<{ data: MonthlySummaryItem[] }>(`/api/accounts/${accountId}/stats/monthly-summary`, { params: range })
      .then((r) => r.data.data),

  byCategory: (accountId: number, range: DateRange = {}) =>
    apiClient
      .get<{ data: CategoryBreakdown[] }>(`/api/accounts/${accountId}/stats/by-category`, { params: range })
      .then((r) => r.data.data),

  byArea: (accountId: number, range: DateRange = {}) =>
    apiClient
      .get<{ data: AreaBreakdown[] }>(`/api/accounts/${accountId}/stats/by-area`, { params: range })
      .then((r) => r.data.data),

  monthlyByCategory: (accountId: number, range: DateRange = {}) =>
    apiClient
      .get<{ data: MonthlyCategoryItem[] }>(`/api/accounts/${accountId}/stats/monthly-by-category`, { params: range })
      .then((r) => r.data.data),

  monthlyByArea: (accountId: number, range: DateRange = {}) =>
    apiClient
      .get<{ data: MonthlyAreaItem[] }>(`/api/accounts/${accountId}/stats/monthly-by-area`, { params: range })
      .then((r) => r.data.data),

  topComercios: (accountId: number, range: DateRange = {}, limit = 10) =>
    apiClient
      .get<{ data: TopComercioItem[] }>(`/api/accounts/${accountId}/stats/top-comercios`, {
        params: { ...range, limit },
      })
      .then((r) => r.data.data),

  kpiTransactions: (accountId: number, kpiId: number, range: DateRange = {}) =>
    apiClient
      .get<{ data: Transaction[] }>(`/api/accounts/${accountId}/kpis/${kpiId}/transactions`, { params: range })
      .then((r) => r.data.data),
}
